src/lib/comms/include/sol-socket.h

```diff
@@ -0,0 +1,352 @@
+     */
+    ssize_t (*recvmsg)(struct sol_socket *s, void *buf, size_t len,
+        struct sol_network_link_addr *cliaddr);
+
+    /**
+     * @brief Function to be called to write data in the socket.
+     */
+    int (*sendmsg)(struct sol_socket *s, const void *buf, size_t len,
+        const struct sol_network_link_addr *cliaddr);
+
+    /**
+     * @brief Function to be called to join a multicast group.
+     */
+    int (*join_group)(struct sol_socket *s, int ifindex,
+        const struct sol_network_link_addr *group);
+
+    /**
+     * @brief Function to be called to bind the socket to a network address.
+     */
+    int (*bind)(struct sol_socket *s, const struct sol_network_link_addr *addr);
+
+    /**
+     * @brief Function to be called to set a option in the socket.
+     */
+    int (*setsockopt)(struct sol_socket *s, enum sol_socket_level level,
+        enum sol_socket_option optname, const void *optval, size_t optlen);
+
+    /**
+     * @brief Function to be called to get a socket option.
+     */
+    int (*getsockopt)(struct sol_socket *s, enum sol_socket_level level,
+        enum sol_socket_option optname, void *optval, size_t *optlen);
+};
+
+#ifndef SOL_NO_API_VERSION
+/**
+ * @brief Macro used to check if a struct @ref sol_socket_type
+ * has the expected API version.
+ *
+ * In case it's a wrong version, it'll return extra arguments passed
+ * to the macro.
+ */
+#define SOL_SOCKET_TYPE_CHECK_API_VERSION(type_, ...) \
+    if (SOL_UNLIKELY((type_)->api_version != \
+        SOL_SOCKET_TYPE_API_VERSION)) { \
+        SOL_ERR("Unexpected API version (response is %u, expected %u)", \
+            (type_)->api_version, SOL_SOCKET_TYPE_API_VERSION); \
+        return __VA_ARGS__; \
+    }
+#else
+#define SOL_SOCKET_TYPE_CHECK_API_VERSION(type_, ...)
+#endif
+
+/**
+ * @brief Structure to represent a socket.
+ *
+ * @see sol_socket_type
+ */
+struct sol_socket {
+#ifndef SOL_NO_API_VERSION
+#define SOL_SOCKET_API_VERSION (1)  /**< compile time API version to be checked during runtime */
+    /**
+     * must match #SOL_SOCKET_API_VERSION at runtime.
+     */
+    uint16_t api_version;
+#endif
+    const struct sol_socket_type *impl;
+};
+
+#ifndef SOL_NO_API_VERSION
+/**
+ * @brief Macro used to check if a struct @ref sol_socket
+ * has the expected API version.
+ *
+ * In case it's a wrong version, it'll return extra arguments passed
+ * to the macro.
+ */
+#define SOL_SOCKET_CHECK_API_VERSION(sock_, ...) \
+    if (SOL_UNLIKELY((sock_)->api_version != \
+        SOL_SOCKET_API_VERSION)) { \
+        SOL_ERR("Unexpected API version (response is %u, expected %u)", \
+            (sock_)->api_version, SOL_SOCKET_API_VERSION); \
+        return __VA_ARGS__; \
+    }
+#else
+#define SOL_SOCKET_CHECK_API_VERSION(sock_, ...)
+#endif
+
+/**
+ * @brief Creates an endpoint for communication.
+ *
+ * This function creates a socket using the system default implementation.
+ *
+ * @param domain A communication domain.
```

Comment:
domain and protocol are bitwise? enums? of what? Please document here what to use in these parameters.

As I talked to you in person, maybe a better approach is to change the `_new()` to take a `_config` like we do for gpio and the others, in there you also give the on_read/on_write callbacks and data. Then instead of "set_on_read(cb,data)", you simple do "monitor_on_read(boolean)"